### PR TITLE
Refactoring RDIP + new features

### LINKED ISSUE #7: Close then open a new TCP connection?
Should it be possible to close the socket connection to the debug server, then open a new connection? 

Currently if I close the socket connection, I get the message `system:10054` from SketchUp in my debugger window. The debug server does not respond when I attempt to open a new connection.

Attached is a screen-shot of a simple session. 
- I start SketchUp with the debug server enabled `sketchup.exe -rdebug "ide"`
- I initialize a `TCPSocket`
- I send a `where` command. SketchUp acknowledges the command.
- I close the connection. SketchUp says `system:10054`
- I am not able to open a new connection.

NOTE - I have also seen the message `asio.misc:2` when disconnecting but I'm not sure how to reproduce one result over the other.

### LINKED ISSUE #10: Race condition in RDIP::Connection with multiple server responses.
When `RDIP::Connection` handles multiple commands that require passing control back to the server thread via `server_response_`, subsequent commands have an opportunity to overwrite the previous ones.

I noticed this behavior while debugging `vscode-ruby` to help investigate rubyide/vscode-ruby#95.  When using PuTTY to connect to the debugger and issue commands directly, I never experience a problem, most likely due to the fact that manually typing commands takes a significantly longer time compared to an automated process.

When `vscode-ruby` processes a breakpoint or step, it submits four requests: `thread list`, `where`, `var local`, and `var global`.  Of note, the two variable requests are sent immediately one after the other, without waiting for a response between.  It is these two requests that `RDIP::Connection` needs to forward to the server thread.  And it would seem that, on my computer at least, it is quite often the case that the `var local` command gets overwritten before the server thread can handle it.  This results in the debugger only sending back the global variables.

For reference, if I modify `vscode-ruby` to wait for the local variables before requesting the globals, I do not experience the issue.

The best solution likely is to use a queue of some form.  That said, something as simple as the following probably could work in the various locations the pattern appears:

```cpp
while (server_response_) { /* Do nothing */ }
server_response_ = std::bind( ... );
process_server_response_ = std::bind( ... );
server_wait_cond_.notify_all();
```

### PR DESCRIPTION
**Pull request by seraku24**
__*Whew!*__ I believe this is my biggest contribution to any github project to date.  Here's hoping that it works for folks.  `(:`

While the commit message lists all of the changes, here are a few highlights:

### Proper handling of cross-thread work.  (Fixes #10.)
The coordination between the main debugger thread and the I/O service thread has been improved.

### Improved attach/detach logic.  (Fixes #7.)
IDEs can attach to SketchUp at any time, even after stopping a previous debugging session.  There is better handling of errors, allowing the SketchUp process to resume execution if the connection to the IDE is lost.  Additionally, the SketchUp process no longer hangs at launch by default.

To enable debugging of scripts that run automatically (i.e. extensions), a new command-line argument, `wait`, may optionally be specified to force SketchUp to wait until a debugger client has attached.

`SketchUp -rdebug "ide port=1234 wait"`

### Conditional breakpoints supported.
This was a relatively easy item to add, building on top of the code for evaluating arbitrary expressions.

NOTE: The current version of `vscode-ruby` only has preliminary support for conditional breakpoints, and as such it requires a few code changes to fully enable it.

---

@thomthom, @BugraB When you are able, I would greatly appreciate your time and help reviewing the code as well as testing the proposed changes on different systems.

### CODE REVIEW

**Comment by BugraB** [comment]
Whew indeed! I will take a look but we are a bit busy so it might take a while to review and test it.
Thank you!

File: Build/postbuild.bat [hunk 1]
```diff
@@ -1,4 +1,35 @@
-rem Comment these out to post-build copy the DLL to SU installation directory.
-rem set SUDIR=C:\Program Files (x86)\SketchUp\SketchUp 2014
-rem xcopy /dy %TARGETDIR%SURubyDebugger.dll %SUDIR%
-rem xcopy /dy %TARGETDIR%SURubyDebugger.pdb %SUDIR%
+@ECHO OFF
```
Comment (seraku24) [inline Build/postbuild.bat]: Done.  The default behavior should be to skip deployment.

**Comment by seraku24** [comment]
> However, we will have to update the project readme to indicate this important change.

I will take a look at updating `README.md` and get a commit pushed.  On a related note, were you able to confirm conditional breakpoints worked in RubyMine?  If so, then it can be removed from the "to do" list while we're at it.

> Also, you mentioned that it is now possible disconnect and reconnect the debugger. I wasn't able to reconnect in RubyMine after connecting. Have you tested this in VS Code? This is not a big deal since it doesn't make things any worse but it was mentioned in your PR description.

Well, that is unfortunate.  But it is hardly unexpected as I have only been able to test using VS Code and PuTTY.  I was worried that other IDEs may not follow the protocol the same way and therefore experience odd behavior.  I had looked into getting a free license for RubyMine for test purposes, but it appears to only be available to actual project members and not an outside contributor as I would be.

Hopefully, it should be possible for me to reproduce the behavior on my side using PuTTY to manually issue the commands as RubyMine sends them.  If you could please post the debug output from the SketchUp process, that should include all of the commands received as well as any error messages if it happens there is a problem at the socket level.

File: README.md [hunk 2]
```diff
@@ -28,17 +34,23 @@ curl https://github.com/SketchUp/sketchup-ruby-debugger/releases/download/1.0.3.
...
 
-Notes:
-- The port should match the remote debugger port setting configured in the IDE. Default port is 1234.
-- SketchUp will start up and appear to be frozen. It is waiting for the debugger to show up.
-- Launch remote debugging in the IDE, SketchUp should continue running. You should see breakpoints hit when Ruby code execution reaches the specified lines.
+## Command-line arguments:
+When launching SketchUp with debugging enabled, the following command-line arguments are supported:
+
+```
+-rdebug "ide [port=<number>] [wait]"
```
Comment (thomthom) [inline README.md]: Nice! I will update the Wiki page for RubyMine once this is merged and I've had some time to play with it.

The ability to disconnected sounded interesting. Will have to set this up with VSCode and see how that works.

**Comment by seraku24** [comment]
@BugraB, @thomthom, just a heads up.  I added support for the `pause`/`interrupt` commands from the `ruby-debug-ide` protocol.  This simply instructs the debugger to stop at the earliest opportunity, which can be useful if you need to break into a long-running script but do not want to set an explicit breakpoint.  If no Ruby code is currently running in SketchUp, then the request simply hangs around until the next function is executed.

File: DebugServer/Log.h [hunk 3]
```diff
@@ -1,17 +1,81 @@
...
 
-inline void Log(const char* str) {
+namespace Log {
+
+static inline std::string Context(const char *filename, size_t line, const char *function) {
+  std::ostringstream output;
+
+  static auto start_time = std::chrono::high_resolution_clock::now();
+  auto current_time = std::chrono::high_resolution_clock::now();
+  auto elapsed = std::chrono::duration_cast<std::chrono::microseconds>(current_time - start_time).count() * 0.000001;
```
Comment (BugraB) [inline DebugServer/Log.h]: We internally follow Google's C++ style guide but we haven't adhered to that very strictly in this project (It's just been a pet project for a few of us). But we generally wrap lines at 80 characters. Not a big deal for this code but we might want to do a pass later to fix the coding style.

https://google.github.io/styleguide/cppguide.html

**Comment by seraku24** [comment]
> This is looking very good! I looked into the detach-reattach issue with RubyMine and it does get a detach command, but doesn't re-attach. Not sure why.

Oh, are you actually seeing the `detach` command or are you seeing `quit` or `exit`?  VS Code sends a `quit` command when you stop debugging.  But I just checked the `ruby-debug-ide` project and see that `detach` can occur for "remote debugging (or local attach)".  The reason it is not working, then, is that I did not support that command.  It's an easy fix to add that as an alternative for `quit` and `exit`, since the expected behavior is the same as far as `sketchup-ruby-debugger` is concerned.  Try this update to the regex on your side to see if it works:

```c++
static const std::regex quit_regex("^(?:q(?:uit)?|exit|detach)$", std::regex_constants::icase);
```

> Also, I couldn't get Pause to work in RubyMine. I see the pause commands come through in the log, but the execution continues. I haven't looked into why.

Pausing is tricky because it only shows immediate results if there is Ruby code running in SketchUp.  Can you confirm that was the case?  For reference, here is what I used when testing it:

```ruby
def pause_test
  keep_looping = true
  iterations = 0
  while keep_looping do
    iterations += 1
  end
  return iterations
end
```

Load that in SketchUp and then run `pause_test` from the Ruby console.  The process should hang due to the (effectively) infinite loop.  Press "pause" in your IDE and you should see execution stop on the `iterations += 1` line.  You can then use the "eval" feature of the IDE to change `keep_looping = false` which should allow the function to return when you continue execution.

**NOTE:** In both cases here, if you are still encountering problems, let's open an issue for each one so we can track things outside of this PR.

**Comment by BugraB** [comment]
> Oh, are you actually seeing the detach command or are you seeing quit or exit?

I am just seeing detach and nothing else. Your suggested change to the regex did the trick! I'll push that change momentarily.

**Comment by BugraB** [comment]
My pause test was very similar to yours. I just tried your exact code and hitting pause repeatedly in RubyMine appears ineffective. It seems to send a "pause 1" message rather than a "pause", so the regex does not match it. I haven't looked up what that 1 means at the end. Do you have a good resource documenting the ruby-debug-ide protocol? When we were first writing this code, we couldn't find a real complete documentation of the protocol so we were mostly operating in trial-and-error mode :)

**Comment by seraku24** [comment]
Ah, that's my bad.  I didn't double-check the protocol reference to see that `pause` (unlike `interrupt`) is supposed to accept a thread number after it, controlling which thread you want to pause.  VS Code only sends `pause` on its own, which is technically a bug with `vscode-ruby`.  But then again, it doesn't properly handle threading with Ruby.

For your information, the protocol reference is at https://github.com/ruby-debug/ruby-debug-ide.  As the documentation is quite out-of-date, the project members simply recommend looking at the code base directly.  For example, here are all of the commands in the protocol:

https://github.com/ruby-debug/ruby-debug-ide/tree/master/lib/ruby-debug-ide/commands

Let's update the regex to the following:

```c++
static const std::regex pause_regex("^(?:pause(?:\\s+(\\d+))?|i(?:nterrupt)?)$", std::regex_constants::icase);
```

This will capture the thread number, if provided, although for now we can safely ignore it.

**Comment by BugraB** [comment]
Not sure how I missed that ruby-debug-ide repo, thank you.
And I applied the suggested change on pause_regex, works well in RubyMine.

I'll be out for the holidays so I may be slow to respond. Thanks again for working on this!

**Comment by seraku24** [comment]
That's great to hear things are all working with RubyMine.  Hope you have a good holiday!  `(:`

**Pull request merged**